src/object/obj_verify.c

```diff
@@ -0,0 +1,672 @@
+			       shard_b, cur_b->iod.iod_recxs->rx_idx);
+			return -DER_MISMATCH;
+		}
+
+		if (cur_a->iod.iod_recxs->rx_nr !=
+		    cur_b->iod.iod_recxs->rx_nr) {
+			D_INFO(DF_OID" (reps %u, inconsistent) "
+			       "shard %u has EV rec len %lu, "
+			       "but shard %u has EV rec len %lu.\n",
+			       DP_OID(oid), reps, shard_a,
+			       cur_a->iod.iod_recxs->rx_nr,
+			       shard_b, cur_b->iod.iod_recxs->rx_nr);
+			return -DER_MISMATCH;
+		}
+
+		/* Fall through. */
+	case VOS_ITER_SINGLE:
+		if (cur_a->iod.iod_size != cur_b->iod.iod_size) {
+			D_INFO(DF_OID" (reps %u, inconsistent) "
+			       "type %u, shard %u has rec size %lu, "
+			       "but shard %u has rec size %lu.\n",
+			       DP_OID(oid), reps, cur_a->type, shard_a,
+			       cur_a->iod.iod_size, shard_b,
+			       cur_b->iod.iod_size);
+			return -DER_MISMATCH;
+		}
+
+		/* Punched record, do nothing. */
+		if (cur_a->iod.iod_size == 0)
+			break;
+
+		D_ASSERT(cur_a->iod.iod_size != DAOS_SIZE_MAX);
+
+		rc = dc_obj_verify_fetch(dova_a);
+		if (rc != 0)
+			return rc;
+
+		rc = dc_obj_verify_fetch(dova_b);
+		if (rc != 0)
+			return rc;
+
+		D_ASSERT(dova_a->fetch_iov.iov_buf == dova_a->fetch_buf);
+		D_ASSERT(dova_b->fetch_iov.iov_buf == dova_b->fetch_buf);
+
+		if (dova_a->fetch_iov.iov_len != dova_b->fetch_iov.iov_len) {
+			D_INFO(DF_OID" (reps %u, inconsistent) "
+			       "type %u, fetched %ld bytes from shard %u, "
+			       "but fetched %ld bytes from shard %u.\n",
+			       DP_OID(oid), reps, cur_a->type,
+			       dova_a->fetch_iov.iov_len, shard_a,
+			       dova_b->fetch_iov.iov_len, shard_b);
+			return -DER_MISMATCH;
+		}
+
+		if (memcmp(dova_a->fetch_iov.iov_buf, dova_b->fetch_iov.iov_buf,
+			   dova_a->fetch_iov.iov_len) != 0) {
+			D_INFO(DF_OID" (reps %u, inconsistent) "
+			       "type %u, shard %u and shard %u have "
+			       "different data, size %lu.\n",
+			       DP_OID(oid), reps, cur_a->type, shard_a, shard_b,
+			       dova_a->fetch_iov.iov_len);
+			return -DER_MISMATCH;
+		}
+
+		break;
+	default:
+		D_ASSERT(0);
+	}
+
+	return 0;
+}
+
+int
+dc_obj_verify_rdg(struct dc_object *obj, struct dc_obj_verify_args *dova,
+		  uint32_t rdg_idx, uint32_t reps, daos_epoch_t epoch)
+{
+	daos_obj_id_t	oid = obj->cob_md.omd_id;
+	daos_handle_t	th;
+	uint32_t	start = rdg_idx * reps;
+	int		rc = 0;
+	int		i;
+
+	rc = dc_tx_local_open(obj->cob_coh, epoch, &th);
+	if (rc != 0)
+		return rc;
+
+	for (i = 0; i < reps; i++) {
+		struct dc_obj_verify_cursor	*cursor = &dova[i].cursor;
+
+		memset(&dova[i].anchor, 0, sizeof(dova[i].anchor));
+		memset(&dova[i].dkey_anchor, 0, sizeof(dova[i].dkey_anchor));
+		memset(&dova[i].akey_anchor, 0, sizeof(dova[i].akey_anchor));
+		dc_obj_shard2anchor(&dova[i].dkey_anchor, start + i);
+
+		dova[i].th = th;
+		dova[i].eof = 0;
+		dova[i].non_exist = 0;
+
+		cursor->gen = 0;
+		cursor->ptr = NULL;
+		dc_obj_verify_reset_cursor(cursor);
+
+		rc = dc_obj_verify_list(&dova[i]);
```

Comment:
do we expect to enumerate and cache all keys of this RDG in this function?

Reply:
No, it is just does the first enumeration for every replica.